### PR TITLE
[SPARK-47776][SS] Disallow binary inequality collation be used in key schema of stateful operator

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to disallow using binary inequality collation column in the key schema of stateful operator. Worth noting that changing the collation for the same string column during the query restart was already disallowed at the time of introduction of collation.

### Why are the changes needed?

state store API is heavily relying on the fact that provider implementation performs O(1)-like get and put operation. While the actual implementation would be dependent on the state store provider, it is intuitive to assume that these providers only do lookup of the key based on binary format (implying binary equality).

That said, even though the column spec is case insensitive, state store API wouldn't take this into consideration, and could lead to produce the wrong result. e.g. Determiniing 'a' and 'A' differently while the column is case insensitive.

### Does this PR introduce _any_ user-facing change?

No, as it wasn't released yet.

### How was this patch tested?

New UTs.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan @dbatomic @anishshri-db

**Comment by cloud-fan** [comment]
We can have O(1) lookup for string with collation values, using a sort key. WIP PR: https://github.com/apache/spark/pull/45929

**Comment by HeartSaVioR** [comment]
What about disallowing this first? It's not only streaming aggregation. We will need to do the same approach for every stateful operator. Before we see the clear path forward, let's prevent the correctness issue for now.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 1]
```diff
@@ -635,6 +635,22 @@ object StateStore extends Logging {
     storeProvider.getStore(version)
   }
 
+  private def disallowBinaryInequalityColumn(schema: StructType): Unit = {
```
Comment (dbatomic) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:638]: You can use following utility function:
```scala
  def isBinaryStable(dataType: DataType): Boolean = !dataType.existsRecursively {
    case st: StringType => !CollationFactory.fetchCollation(st.collationId).supportsBinaryEquality
    case _ => false
  }

```

**Comment by dbatomic** [comment]
> What about disallowing this first? It's not only streaming aggregation. We will need to do the same approach for every stateful operator. Before we see the clear path forward, let's prevent the correctness issue for now.

Yeah, let's go with this change. Once hash agg support change goes in we can try to enable the streaming code. Btw, thank you for adding the test! We will use that as the starting point.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:638]: Thanks, done!

**Review by dbatomic — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
GA only failed in Docker integration which is unrelated.

**Comment by HeartSaVioR** [comment]
Merging to master.

**Pull request closed without merging**